### PR TITLE
Complete XML documentation

### LINKED ISSUE #753: Add Documentation of XML Tags and Attributes
Our [XML reference](https://github.com/precice/precice/wiki/XML-Reference) shows that many tags and attributes are not documented. This state devalues the reference and should be fixed.

# Undocumented Tags

* sink
* absolute-convergence-measure
* aitken
* broyden
* constant
* data
* exchange
* initial-relaxation
* IQN-ILS
* IQN-IMVJ
* max-iterations
* max-time
* max-time-windows
* max-used-iterations
* min-iteration-convergence-measure
* nearest-neighbor
* nearest-projection
* participant
* participants
* preconditioner
* rbf-compact-polynomial-c0
* rbf-compact-polynomial-c6
* rbf-compact-tps-c2
* rbf-gaussian
* rbf-inverse-multiquadrics
* rbf-multiquadrics
* rbf-thin-plate-splines
* rbf-volume-splines
* relative-convergence-measure
* relaxation
* residual-relative-convergence-measure
* time-window-size
* time-windows-reused

# Undocumented Attributes

Tag | Attribute
--- | ---
absolute-convergence-measure | limit
data | mesh
data | name
exchange | data
exchange | from
exchange | initialize
exchange | mesh
exchange | to
extrapolation-order | value
filter | limit
filter | type
imvj-restart-mode | type
initial-relaxation | enforce
initial-relaxation | value
max-iterations | value
max-time | value
max-time-windows | value
max-used-iterations | value
min-iteration-convergence-measure | min-iterations
nearest-neighbor | constraint
nearest-neighbor | direction
nearest-neighbor | from
nearest-neighbor | timing
nearest-neighbor | to
nearest-projection | constraint
nearest-projection | direction
nearest-projection | from
nearest-projection | timing
nearest-projection | to
participant | control
participant | name
participants | first
participants | second
path | name
rbf-compact-polynomial-c0 | constraint
rbf-compact-polynomial-c0 | direction
rbf-compact-polynomial-c0 | from
rbf-compact-polynomial-c0 | timing
rbf-compact-polynomial-c0 | to
rbf-compact-polynomial-c6 | constraint
rbf-compact-polynomial-c6 | direction
rbf-compact-polynomial-c6 | from
rbf-compact-polynomial-c6 | timing
rbf-compact-polynomial-c6 | to
rbf-compact-tps-c2 | constraint
rbf-compact-tps-c2 | direction
rbf-compact-tps-c2 | from
rbf-compact-tps-c2 | timing
rbf-compact-tps-c2 | to
rbf-gaussian | constraint
rbf-gaussian | direction
rbf-gaussian | from
rbf-gaussian | timing
rbf-gaussian | to
rbf-inverse-multiquadrics | constraint
rbf-inverse-multiquadrics | direction
rbf-inverse-multiquadrics | from
rbf-inverse-multiquadrics | timing
rbf-inverse-multiquadrics | to
rbf-multiquadrics | constraint
rbf-multiquadrics | direction
rbf-multiquadrics | from
rbf-multiquadrics | timing
rbf-multiquadrics | to
rbf-thin-plate-splines | constraint
rbf-thin-plate-splines | direction
rbf-thin-plate-splines | from
rbf-thin-plate-splines | timing
rbf-thin-plate-splines | to
rbf-volume-splines | constraint
rbf-volume-splines | direction
rbf-volume-splines | from
rbf-volume-splines | timing
rbf-volume-splines | to
relative-convergence-measure | limit
relaxation | value
residual-relative-convergence-measure | limit
time-window-size | method
time-window-size | valid-digits
time-window-size | value
time-windows-reused | value

### PR DESCRIPTION
**Pull request by fsimonis**
**List the core changes of this PR**

Add the documentation off missing XML tags and attributes.

**Motivation**

Provide a complete XML reference for the website.

**Additional Information**

Closes #753 

**Reviewers**

Please check if the added documentation is correct.

### CODE REVIEW

File: src/acceleration/config/AccelerationConfiguration.cpp [hunk 1]
```diff
@@ -87,25 +87,29 @@ void AccelerationConfiguration::connectTags(xml::XMLTag &parent)
...
+  std::vector<XMLTag> tags;
   {
     XMLTag tag(*this, VALUE_CONSTANT, occ, TAG);
+    tag.setDocumentation("Accelerates coupling data with constant underrelaxation.");
     addTypeSpecificSubtags(tag);
     tags.push_back(tag);
   }
   {
     XMLTag tag(*this, VALUE_AITKEN, occ, TAG);
+    tag.setDocumentation("Accelerates coupling data with dynamic underrelaxation.");
```
Comment (KyleDavisSA) [inline src/acceleration/config/AccelerationConfiguration.cpp:100]: Should we mention it is Aitken relaxation? 
"Accelerates coupling data with dynamic Aitken under-relaxation."

File: src/acceleration/config/AccelerationConfiguration.cpp [hunk 2]
```diff
@@ -385,36 +405,38 @@ void AccelerationConfiguration::addTypeSpecificSubtags(
...
     XMLTag tagPreconditioner(*this, TAG_PRECONDITIONER, XMLTag::OCCUR_NOT_OR_ONCE);
-    auto   attrPreconditionerType = XMLAttribute<std::string>(ATTR_TYPE)
-                                      .setDocumentation(
-                                          "To improve the performance of a parallel or a multi coupling schemes a preconditioner"
-                                          " can be applied. A constant preconditioner scales every acceleration data by a constant value, which you can define as"
-                                          " an attribute of data. "
-                                          " A value preconditioner scales every acceleration data by the norm of the data in the previous time window."
-                                          " A residual preconditioner scales every acceleration data by the current residual."
-                                          " A residual-sum preconditioner scales every acceleration data by the sum of the residuals from the current time window.")
+    tagPreconditioner.setDocumentation("To improve the performance of a parallel or a multi coupling schemes a preconditioner"
```
Comment (KyleDavisSA) [inline src/acceleration/config/AccelerationConfiguration.cpp:428]: "coupling schemes, a preconditioner"

File: src/cplscheme/config/CouplingSchemeConfiguration.cpp [hunk 3]
```diff
@@ -455,20 +446,27 @@ void CouplingSchemeConfiguration::addTransientLimitTags(
     xml::XMLTag &      tag)
 {
   using namespace xml;
-  XMLTag               tagMaxTime(*this, TAG_MAX_TIME, XMLTag::OCCUR_NOT_OR_ONCE);
+  XMLTag tagMaxTime(*this, TAG_MAX_TIME, XMLTag::OCCUR_NOT_OR_ONCE);
+  tagMaxTime.setDocumentation("Defined the end of the simulation as total time.");
+
   XMLAttribute<double> attrValueMaxTime(ATTR_VALUE);
+  attrValueMaxTime.setDocumentation("The value of the maximum.");
```
Comment (KyleDavisSA) [inline src/cplscheme/config/CouplingSchemeConfiguration.cpp:453]: "The value of the maximum simulation time" ?

File: src/cplscheme/config/CouplingSchemeConfiguration.cpp [hunk 4]
```diff
@@ -477,7 +475,7 @@ void CouplingSchemeConfiguration::addTransientLimitTags(
   } else {
     allowedMethods = {VALUE_FIXED};
   }
-  auto attrMethod = makeXMLAttribute(ATTR_METHOD, VALUE_FIXED).setOptions(allowedMethods);
+  auto attrMethod = makeXMLAttribute(ATTR_METHOD, VALUE_FIXED).setOptions(allowedMethods).setDocumentation("Is the time window size fixed or does the first participant choose it?");
```
Comment (KyleDavisSA) [inline src/cplscheme/config/CouplingSchemeConfiguration.cpp:478]: Should there be a question in the documentation?

File: src/cplscheme/config/CouplingSchemeConfiguration.cpp [hunk 5]
```diff
@@ -531,6 +535,7 @@ void CouplingSchemeConfiguration::addTagAbsoluteConvergenceMeasure(
   XMLTag tagConvergenceMeasure(*this, TAG_ABS_CONV_MEASURE, XMLTag::OCCUR_ARBITRARY);
   addBaseAttributesTagConvergenceMeasure(tagConvergenceMeasure);
   XMLAttribute<double> attrLimit(ATTR_LIMIT);
+  attrLimit.setDocumentation("Limit under which the measure is considered to have converged.");
```
Comment (KyleDavisSA) [inline src/cplscheme/config/CouplingSchemeConfiguration.cpp:538]: Perhaps a better description of each convergence measure type? From the absoluteConvergenceMeasure.hpp

 " The convergence is evaluated by looking at the two norm of the differences between each data value from the new and old data set. If the two norm is equal or below a given limit, convergence is achieved."

We can add this for the others below?

File: src/cplscheme/config/CouplingSchemeConfiguration.cpp [hunk 6]
```diff
@@ -531,6 +535,7 @@ void CouplingSchemeConfiguration::addTagAbsoluteConvergenceMeasure(
   XMLTag tagConvergenceMeasure(*this, TAG_ABS_CONV_MEASURE, XMLTag::OCCUR_ARBITRARY);
```
Comment (KyleDavisSA) [inline src/cplscheme/config/CouplingSchemeConfiguration.cpp:535]: There is no documentation for the convegence measure type? Only for the value

File: src/cplscheme/config/CouplingSchemeConfiguration.cpp [hunk 7]
```diff
@@ -591,8 +598,10 @@ void CouplingSchemeConfiguration::addTagMaxIterations(
     xml::XMLTag &tag)
 {
   using namespace xml;
-  XMLTag            tagMaxIterations(*this, TAG_MAX_ITERATIONS, XMLTag::OCCUR_NOT_OR_ONCE);
+  XMLTag tagMaxIterations(*this, TAG_MAX_ITERATIONS, XMLTag::OCCUR_NOT_OR_ONCE);
+  tagMaxIterations.setDocumentation("Allows to specify a maximum amount of iterations. The simulation will end if this limit is exceeded.");
```
Comment (KyleDavisSA) [inline src/cplscheme/config/CouplingSchemeConfiguration.cpp:602]: Is this not maximum iterations until the time-window is incremented, regardless if convergence is reached or not?

File: src/mapping/config/MappingConfiguration.cpp [hunk 8]
```diff
@@ -106,23 +114,32 @@ MappingConfiguration::MappingConfiguration(
...
-  XMLAttribute<std::string> attrFromMesh(ATTR_FROM);
-  XMLAttribute<std::string> attrToMesh(ATTR_TO);
+  auto attrFromMesh = XMLAttribute<std::string>(ATTR_FROM)
+                          .setDocumentation("The mesh to map the data from.");
+
+  auto attrToMesh = XMLAttribute<std::string>(ATTR_TO)
+                        .setDocumentation("The mesh to map the data to.");
 
   auto attrConstraint = XMLAttribute<std::string>(ATTR_CONSTRAINT)
+                            .setDocumentation("Use consistent for ")
```
Comment (KyleDavisSA) [inline src/mapping/config/MappingConfiguration.cpp:138]: I think there is something missing?

File: src/cplscheme/config/CouplingSchemeConfiguration.cpp (hunk 7 above)
Comment (fsimonis) [reply src/cplscheme/config/CouplingSchemeConfiguration.cpp:602]: @BenjaminRueth can you enlighten us?

File: src/cplscheme/config/CouplingSchemeConfiguration.cpp (hunk 4 above)
Comment (fsimonis) [reply src/cplscheme/config/CouplingSchemeConfiguration.cpp:478]: Good point! I reworded this doc string.

File: src/cplscheme/config/CouplingSchemeConfiguration.cpp (hunk 7 above)
Comment (uekerman) [reply src/cplscheme/config/CouplingSchemeConfiguration.cpp:617]: It's the iterations per time window. If reached we go to the next time window.

**Comment by fsimonis** [comment]
The result of this PR can be seen here https://www.precice.org/precice.github.io_future/configuration-xml-reference.html

**Pull request merged**